src/net/stackerdb/db.rs

```diff
@@ -0,0 +1,602 @@
+        write_time INTEGER NOT NULL,
+        
+        PRIMARY KEY(rc_consensus_hash,chunk_id)
+    );
+    "#;
+
+const CHUNKS_DB_INDEX_BODIES: &'static [(&'static str, &'static str)] =
+    &[("chunk_versions", "(rc_consensus_hash,chunk_id,version)")];
+
+pub const NO_VERSION: i64 = 0;
+
+/// Private struct for loading the data we need to validate an incoming chunk
+pub struct ChunkValidation {
+    pub stacker: StacksAddress,
+    pub version: u32,
+    pub write_time: u64,
+}
+
+impl FromRow<ChunkMetadata> for ChunkMetadata {
+    fn from_row<'a>(row: &'a Row) -> Result<ChunkMetadata, db_error> {
+        let rc_consensus_hash: ConsensusHash =
+            ConsensusHash::from_column(row, "rc_consensus_hash")?;
+        let chunk_id: u32 = row.get_unwrap("chunk_id");
+        let chunk_version: u32 = row.get_unwrap("version");
+        let data_hash_str: String = row.get_unwrap("data_hash");
+        let data_hash =
+            Sha512Trunc256Sum::from_hex(&data_hash_str).map_err(|_| db_error::ParseError)?;
+        let message_sig_str: String = row.get_unwrap("signature");
+        let signature =
+            MessageSignature::from_hex(&message_sig_str).map_err(|_| db_error::ParseError)?;
+
+        Ok(ChunkMetadata {
+            rc_consensus_hash,
+            chunk_id,
+            chunk_version,
+            data_hash,
+            signature,
+        })
+    }
+}
+
+impl FromRow<ChunkValidation> for ChunkValidation {
+    fn from_row<'a>(row: &'a Row) -> Result<ChunkValidation, db_error> {
+        let stacker = StacksAddress::from_column(row, "stacker")?;
+        let version: u32 = row.get_unwrap("version");
+        let write_time_i64: i64 = row.get_unwrap("write_time");
+        if write_time_i64 < 0 {
+            return Err(db_error::ParseError);
+        }
+        let write_time = write_time_i64 as u64;
+
+        Ok(ChunkValidation {
+            stacker,
+            version,
+            write_time,
+        })
+    }
+}
+
+impl FromRow<StackerDBChunkData> for StackerDBChunkData {
+    fn from_row<'a>(row: &'a Row) -> Result<StackerDBChunkData, db_error> {
+        let chunk_id: u32 = row.get_unwrap("chunk_id");
+        let chunk_version: u32 = row.get_unwrap("version");
+        let data: Vec<u8> = row.get_unwrap("data");
+        let message_sig_str: String = row.get_unwrap("signature");
+        let sig = MessageSignature::from_hex(&message_sig_str).map_err(|_| db_error::ParseError)?;
+
+        Ok(StackerDBChunkData {
+            chunk_id,
+            chunk_version,
+            sig,
+            data,
+        })
+    }
+}
+
+/// Convert a smart contract name to a table name
+pub fn stackerdb_table(addr: &ContractId) -> String {
+    let normalized_name = format!("{}", &addr)
+        .replace("-", "\\x2d")
+        .replace("_", "\\x5f")
+        .replace(".", "_");
+
+    format!("stackerdb_{}", normalized_name)
+}
+
+/// Load up chunk metadata from the database, given the primary key.
```

Comment:
```suggestion
/// Load a chunk's metadata from the database, keyed by the chunk's smart contract, the reward cycle consensus hash, and the chunk identifier.
```

Looking at this PR after #3551 and #3552, I think there's a naming conflict with "StackerDB" -- you refer to the local database as "the `StackerDB`", but also, throughout the code, different smart contracts' associated chunk storage is each called "a `StackerDB`". This is already kind of confusing to me, and this is pretty early on in the life of this code, so I think we'd probably be better served with a different way to talk about these things.

Reply:
I'm having a hard time understanding what the ask here is.  The "smart contracts' associated chunk storage" _is_ a stacker DB.  The node maintains a local replica of that state for each stacker DB it subscribes to.  But, I don't see the value in calling the local replica by a different name, since I think that's already understood simply because a single process cannot be a whole p2p network.

Reply:
I think you're referring to the fact that `StackerDB` the struct currently stores all stacker DB chunks for all subscribed stacker DBs?  If so, then I'll rename `StackerDB` to `StackerDBSet` to disambiguate it.